### ISSUE TITLE
KeyboardLanguageContext class?

### ISSUE THREAD

**Issue by kendonB** [dictation-toolbox/dragonfly#81]
We have noticed some trouble getting commands working with non- English Language / English Keyboard setups https://github.com/dictation-toolbox/Caster/issues/512.

Is it possible to create something like `AppContext` that will test Language + Keyboard like AppContext tests Apps? 

Reposting the "carat" behavior from the linked issue:
```
German language, German keyboard: ^^
German language, English US keyboard: `
English language, German keyboard: &
English language, English US keyboard: ^
```

**Comment by Danesprite** [comment]
A context class is certainly one way to handle this. Typing characters that require key composition would be hard to get right universally, but an option could be added for it. The problem should also be mentioned in the documentation.

A related issue is what happens if you switch layouts and use dragonfly to type. Windows stores locale information with each thread and doesn't automatically update it when the keyboard layout changes. This is the reason that "&" is being typed in the 3rd case, dragonfly is typing as if the English US layout was active. Sometimes this is desirable, sometimes it isn't. Dragon itself clearly handles layout changes where it can, e.g. when you say something like "press x", so dragonfly probably should too.

I think it should be possible to set an event hook for layout/language changes, although it might be difficult to set up nicely. The `Typeable` objects for  keys would need to be changed/cached on layout switches. Different values will need to be used when calling `windll.user32.GetKeyboardLayout()`. Special keys like `VK_RETURN` or `VK_DELETE` would not need to be updated.

**Comment by Danesprite** [comment]
There are some differences with how keyboard layouts work in Windows 8 and above, so I'm not sure if the problem is really worth fixing if it doesn't affect newer OS versions.

I've done a little research into it anyway. According to [this Stack Overflow answer](https://stackoverflow.com/a/17908270), the [`WM_INPUTLANGCHANGE` message](http://msdn.microsoft.com/en-us/library/windows/desktop/ms632629.aspx) is only sent to the active window, so we can't get at it from dragonfly. I can verify that this is unfortunately true. They mention another solution which can't (to my knowledge) be done with Python.

I've come up with another solution that kind of works:

```Python
import win32api
import win32gui
import win32process

# Get the current window's locale identifier.
handle = win32gui.GetForegroundWindow()
threadId, _ = win32process.GetWindowThreadProcessId(handle)
lang = win32api.GetKeyboardLayout(threadId)

# Use it to get the virtual key code for the caret character.
code = win32api.VkKeyScanEx("^", lang)
print(code, lang)

```

I've tested using the win32 `Keyboard` and `Typeable` classes with this approach. Unfortunately it only seems to work for specific characters. I'm guessing this is something to do with the `VkKeyScanEx` function because `lang` is returning the correct value. I have only been testing with a not up-to-date Windows 7 install, so that issue might not be present in later versions.

**Comment by Danesprite** [comment]
An easier way of consistently typing characters like this is to use the `Text` action's [Unicode keyboard support](https://dragonfly2.readthedocs.io/en/latest/actions.html#windows-unicode-keyboard-support). The `Key` action doesn't support that because modifier keys (e.g. ctrl and alt) are ignored.

@seekM This is a little easier than the solution you came up with in dictation-toolbox/Caster#512.

The solution I mention in the previous comment also doesn't work correctly on Windows 10, so I'm probably missing something about how keyboard input works on Windows.

@Versatilus You seem pretty knowledgeable with this stuff. Any ideas how we might achieve this or is it just a pipe dream?

**Comment by Danesprite** [comment]
A few updates on this issue.

# Improving keyboard input
I've managed to get dragonfly to type into the current window using the correct keys for the current keyboard layout. It wasn't working before because the [`MapVirtualKeyW()`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mapvirtualkeyw) function is used in [sendinput.py](https://github.com/dictation-toolbox/dragonfly/blob/master/dragonfly/actions/sendinput.py) instead of [`MapVirtualKeyExW()`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mapvirtualkeyexw). The second function uses the layout when determining the scan code. I've also had to get the virtual key codes during the `Typeable.event*()` methods instead. This doesn't seem to reduce the performance, but I'll need to do more testing.

I think we can actually solve another problem with this approach: less unknown key errors. Currently if the keyboard layout doesn't support some key (say `"c"`) and the key's `Typeable` object was not initialised with `is_text=True`, then a `ValueError` will be raised:

```Python
ValueError: Unknown char: 'c'

```

This can be solved for most keys by falling back on the Unicode keyboard events when `VkKeyScanEx()` returns `-1`. Modifiers won't work properly this way, but I think it's better than an error being raised.

I haven't got this part working yet because Unicode events are currently tied to the `Text` class.

# KeyboardLanguageContext class

This still leaves the original point of this issue: a `KeyboardLanguageContext` context class to deal with layout differences like the `^` key in English (normal key) vs German (composition key). I still think it's much easier to just use the Unicode keyboard support, however it should be possible to use the current window's locale identifier somehow. I'm just not sure how to get a human-readable string like from it.

**Comment by Danesprite** [comment]
@kendonB The inconsistent behaviour between keyboard layouts described in your original post should be fixed in `master` now. Dragonfly's `Key` and `Text` actions should always type the correct keys for the current foreground window, e.g. `^` will always press the `^` key instead of `&`. I've updated the [documentation](https://dragonfly2.readthedocs.io/en/latest/actions.html#windows-key-support) on that.

As for `KeyboardLanguageContext`, you could use `FuncContext` and `ContextAction` to do this:

```Python
from dragonfly import (ContextAction, FuncContext, MappingRule, Key,
                       Keyboard)

def get_hkl_context(hkl):
    """ Get a FuncContext for matching keyboard layouts. """
    return FuncContext(lambda: Keyboard.get_current_layout() == hkl)

# Keyboard layout contexts.
# HKL values were retrieved via Keyboard.get_current_layout()
german_ctx = get_hkl_context(67568647)
dutch_ctx = get_hkl_context(68355091)

class KeyRule(MappingRule):
    mapping = {
        # Define default and contextual actions for "type caret".
        "type caret": ContextAction(Key("^"), {
            ((german_ctx | dutch_ctx), Key("^, space")),
        })
    }


```

I don't think this should be added into Dragonfly unless `get_hkl_context()` or an equivalent class can take the name shown in the Windows language bar, e.g. `"German (Germany)"`.

**Comment by Danesprite** [comment]
Closing as resolved because dragonfly2 version 0.19.0 fixes most Win32 keyboard-related issues and my solution above is a good enough fix for the other part of the issue. I'll re-open this if I find a way to reliably get at current keyboard & language names.
